### PR TITLE
Fix reconciling split translations from nYNAB import creates orphan transfers

### PR DESCRIPTION
**Pull request by Wizmaster**
Tentative fix for #2436

It is a bit more convoluted than I would like it to be, but nYNAB exports are missing several key fields on subtransactions.

The code will try to match subtransaction transfers with transaction transfers (missing the transfer_transaction_id field, which indicate that the transfer is linked to a subtransaction). Matching is made by ordering arrays of transactions and subtransactions, and trying to find matching data to store the linked ids. Unmatched data (if that should happens) will still be imported and will behave like currently in Actual.

I also streamlined payee import on transactions and subtransactions. This fix a small bug where payee on subtransactions are not imported.

### CODE REVIEW

**Comment by Wizmaster** [comment]
Here is also a test export file with several subtransaction transfers.

[budget-test-transfer-complex.json](https://github.com/actualbudget/actual/files/14735805/budget-test-transfer-complex.json)

File: packages/loot-core/src/server/importers/ynab5.ts [hunk 1]
```diff
@@ -162,6 +162,129 @@ async function importTransactions(
...
+      if (!map.has(key)) {
+        map.set(key, [transaction]);
+      } else {
+        map.get(key).push(transaction);
+      }
+      return map;
+    }, new Map<string, YNAB5.Transaction[]>());
+
+  const orphanSubtransfer = data.subtransactions.filter(
+    subtransaction => subtransaction.transfer_account_id,
```
Comment (twk3) [inline packages/loot-core/src/server/importers/ynab5.ts:185]: Is every subtransaction with a transfer_account_id orphaned?

File: packages/loot-core/src/server/importers/ynab5.ts [hunk 2]
```diff
@@ -162,6 +162,129 @@ async function importTransactions(
...
+  const orphanSubtransferAcctIdByTrxId = data.transactions
+    .filter(transaction => orphanSubtransferTrxId.includes(transaction.id))
+    .map(
+      transaction =>
+        [transaction.id, transaction.account_id] as [string, string],
+    );
+  const orphanSubtransferAcctIdByTrxIdMap = new Map<string, string>(
+    orphanSubtransferAcctIdByTrxId,
+  );
+  const orphanSubtransferDateByTrxId = data.transactions
```
Comment (twk3) [inline packages/loot-core/src/server/importers/ynab5.ts:199]: Can we reduce the number times we run through this transaction list by returning an object that has date and account Id.

And we are already running through date.transactions and data.subtransactions in the for loops above your conditions here. We could be building the loopup maps we need in those for loops to reduce some loops.

**Review by twk3 — COMMENTED** [review]
@Wizmaster I've left some questions. I haven't reviewed the `orphanTransferComparator` method. But I've taken a look at the rest.

File: packages/loot-core/src/server/importers/ynab5.ts (hunk 1 above)
Comment (Wizmaster) [reply packages/loot-core/src/server/importers/ynab5.ts]: Unfortunately yes, subtransactions are missing `transfer_transaction_id`, and the corresponding transactions too.

File: packages/loot-core/src/server/importers/ynab5.ts (hunk 2 above)
Comment (Wizmaster) [reply packages/loot-core/src/server/importers/ynab5.ts]: Sure, I removed the additional loops on date.transactions and data.subtransactions and merged the code with the initial loops.

File: packages/loot-core/src/server/importers/ynab5.ts [hunk 3]
```diff
@@ -152,16 +152,134 @@ async function importTransactions(
...
+        ? a.date
+        : orphanSubtransferDateByTrxIdMap.get(a.transaction_id);
+    const date_b =
+      'date' in b
+        ? b.date
+        : orphanSubtransferDateByTrxIdMap.get(b.transaction_id);
+    const amount_a = 'date' in a ? a.amount : -a.amount;
+    const amount_b = 'date' in b ? b.amount : -b.amount;
+
+    if (date_a > date_b) return 1;
```
Comment (twk3) [inline packages/loot-core/src/server/importers/ynab5.ts:232]: worth a quick inline comment for future readers to outline the priority of comparison, so they don't have to read the returns

File: packages/loot-core/src/server/importers/ynab5.ts [hunk 4]
```diff
@@ -152,16 +152,134 @@ async function importTransactions(
...
+      let transactionIdx = 0;
+      let subtransactionIdx = 0;
+      do {
+        switch (
+          orphanTransferComparator(
+            transactions[transactionIdx],
+            subtransactions[subtransactionIdx],
+          )
+        ) {
+          case 0:
```
Comment (twk3) [inline packages/loot-core/src/server/importers/ynab5.ts:257]: Let's leave a comment for each case to make this quicker to read for future contributors.

Just a quick, one liner about why we are moving on to the next transaction or the next subtransaction

File: packages/loot-core/src/server/importers/ynab5.ts [hunk 5]
```diff
@@ -152,16 +152,134 @@ async function importTransactions(
...
+    const date_a =
+      'date' in a
+        ? a.date
+        : orphanSubtransferDateByTrxIdMap.get(a.transaction_id);
+    const date_b =
+      'date' in b
+        ? b.date
+        : orphanSubtransferDateByTrxIdMap.get(b.transaction_id);
+    const amount_a = 'date' in a ? a.amount : -a.amount;
+    const amount_b = 'date' in b ? b.amount : -b.amount;
```
Comment (twk3) [inline packages/loot-core/src/server/importers/ynab5.ts:230]: I'm not quite following why the conditionally negate here. Could you clarify that in a code comment?

**Review by twk3 — COMMENTED** [review]
@Wizmaster I few requests for code comment clarifications (and one that I'm not quite following)

Apologies, for all the questions, but I and other who will read this later aren't going to have the same knowledge of the ynab output looks like (none of my subtransactions were transfers for mine). And you clearly did a great deep dive into the export for this PR! Much appreciated!

File: packages/loot-core/src/server/importers/ynab5.ts (hunk 3 above)
Comment (Wizmaster) [reply packages/loot-core/src/server/importers/ynab5.ts:243]: Of course! I added comments.

File: packages/loot-core/src/server/importers/ynab5.ts (hunk 4 above)
Comment (Wizmaster) [reply packages/loot-core/src/server/importers/ynab5.ts:270]: Comments added too 😄

File: packages/loot-core/src/server/importers/ynab5.ts (hunk 5 above)
Comment (Wizmaster) [reply packages/loot-core/src/server/importers/ynab5.ts:240]: The conditional negate is to have transactions (having the date property set) and subtransactions (without date) have comparable amount. They should be the inverse of each other, so here subtransactions are reversed to match transactions.
I added the explanation in a comment too.

**Review by twk3 — APPROVED** [review]
Thanks @Wizmaster !

This is resolving the identified problem, and I really appreciate the comments you added.

Our pipelines are having an issue at the moment, (not just your branch), and once we have that resolved I will merge this. Thanks!

**Pull request merged**